### PR TITLE
fix(python): Check that fractions are between 0 and 1 for `pl.Expr.sample`

### LINKED ISSUE #22024: `list.sample` fraction allows floats between 1.0 and 2.0
### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
df = pl.DataFrame([
    pl.Series('a', [
        ["a"], ["eb", "d"],
    ], pl.List(pl.String)),
])
df.select(pl.col.a.list.sample(fraction=1.2))
```

### Log output

```shell
shape: (2, 1)
+-------------+
| a           |
| ---         |
| list[str]   |
+=============+
| ["a"]       |
|-------------|
| ["eb", "d"] |
+-------------+
```

### Issue description

I don't think this is correct.

### Expected behavior

It should be kept between 0.0 and 1.0.

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

### PR DESCRIPTION
**Pull request by alex-gregory-ds**
Closes #22024.

This adds a check to ensure each element of the `fraction` argument given to `pl.Expr.sample` is between 0 and 1. If it isn't, a `ComputeError` is raised.

Note for reviewers: I'm not sure if this is the right error type to use and I'm not sure whether this should have a unit test in rust, python or both.

### CODE REVIEW

**Comment by vrajat** [comment]
I dont know the expected behaviour if fraction is `None` AND `n` is not set. Refer: https://github.com/pola-rs/polars/issues/22024#issuecomment-3538462191
 
 Here is a set of tests I wrote for myself to check its behaviour:
 
 ```
 diff --git a/py-polars/tests/unit/operations/namespaces/list/test_list.py b/py-polars/tests/unit/operations/namespaces/list/test_list.py
index 3c9e5fe2f7..ae762fe6d0 100644
--- a/py-polars/tests/unit/operations/namespaces/list/test_list.py
+++ b/py-polars/tests/unit/operations/namespaces/list/test_list.py
@@ -424,6 +424,23 @@ def test_list_sample() -> None:
     )
     assert_frame_equal(df, expected_df)

+def test_list_sample_fraction_out_of_bounds():
+    s = pl.Series("values", [[1, 2, 3], [4, 5]])
+    # Fraction > 1.0
+    with pytest.raises(ComputeError, match="fraction.*must be between 0 and 1"):
+        s.list.sample(fraction=1.5)
+    # Fraction < 0.0
+    with pytest.raises(ComputeError, match="fraction.*must be between 0 and 1"):
+        s.list.sample(fraction=-0.1)
+    with pytest.raises(ComputeError, match="fraction.*must be between 0 and 1"):
+        s.list.sample(fraction=None)
+
+def test_list_sample_fraction_null():
+    s = pl.Series("values", [[1, 2, 3], [4, 5]])
+    # Fraction is null for the first row, valid for the second
+    fraction = pl.Series("fraction", [None, 1.5])
+    with pytest.raises(ComputeError, match="fraction.*must be between 0 and 1"):
+        s.list.sample(fraction=fraction)
 ```

**Comment by alex-gregory-ds** [comment]
@vrajat, thanks for suggesting those unit tests. After speaking with reviewers on Discord, I've changed the code so it no longer raises a `ComputeError` and therefore have not added your unit tests as you have written them.

However, your tests cover edge cases that I have not considered so it would be useful to get the opinion of a reviewer on this. In particular, @vrajat, you have written tests for negative fractions and `None` fractions. This is how the code currently behaves in both those instances:

```python
>>> import polars as pl
>>> df = pl.DataFrame(pl.Series("a", [[1, 2, 3], [4, 5]]))
>>> pl.col.a.list.sample(fraction=pl.Series([1.0, 0.0]), with_replacement=False)
<Expr ['col("a").list.sample_fraction(…'] at 0x27DC34C7650>
>>> df.select(pl.col.a.list.sample(fraction=pl.Series([1.0, 0.0]), with_replacement=False))
shape: (2, 1)
┌───────────┐
│ a         │
│ ---       │
│ list[i64] │
╞═══════════╡
│ [1, 2, 3] │
│ []        │
└───────────┘
>>> >>> df.select(pl.col.a.list.sample(fraction=pl.Series([1.1, 0.0]), with_replacement=False))
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    df.select(pl.col.a.list.sample(fraction=pl.Series([1.1, 0.0]), with_replacement=False))
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\AlexG\Documents\polars\py-polars\src\polars\dataframe\frame.py", line 10124, in select
    .collect(optimizations=QueryOptFlags._eager())
     ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\AlexG\Documents\polars\py-polars\src\polars\_utils\deprecation.py", line 97, in wrapper
    return function(*args, **kwargs)
  File "C:\Users\AlexG\Documents\polars\py-polars\src\polars\lazyframe\opt_flags.py", line 328, in wrapper
    return function(*args, **kwargs)
  File "C:\Users\AlexG\Documents\polars\py-polars\src\polars\lazyframe\frame.py", line 2425, in collect
    return wrap_df(ldf.collect(engine, callback))
                   ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
polars.exceptions.ShapeError: cannot take a larger sample than the total population when `with_replacement=false`
>>> df.select(pl.col.a.list.sample(fraction=pl.Series([-5, -1]), with_replacement=False))
shape: (2, 1)
┌───────────┐
│ a         │
│ ---       │
│ list[i64] │
╞═══════════╡
│ []        │
│ []        │
└───────────┘
>>> >>> df.select(pl.col.a.list.sample(fraction=pl.Series([None, 1.0]), with_replacement=False))
shape: (2, 1)
┌───────────┐
│ a         │
│ ---       │
│ list[i64] │
╞═══════════╡
│ null      │
│ [4, 5]    │
└───────────┘
```

The case where the fraction is `None` makes sense to me but the cases where the fractions are negative make less sense to me. Should we raise an error when a negative fraction is provided?

**Comment by vrajat** [comment]
I vote for an error. The intent of -ve fractions is unclear to me.

**Comment by alex-gregory-ds** [comment]
@coastalwhite, @ritchie46, if you have time could I please have a review on this?

**Comment by orlp** [comment]
Am I missing something? This doesn't seem to add any validation at all, just change behavior to ceil instead of floor.

**Comment by alex-gregory-ds** [comment]
Initially this did perform a bound check and raise a compute error - see https://github.com/pola-rs/polars/pull/25366/changes/1a4135c5cff2fedbe8025b614cb0df05e3be91a9.

After discussions on the Polars discord, I decided to change it.

The original issue had a similar example to the one below:

```python
import polars as pl
df = pl.DataFrame([
    pl.Series('a', [
        ["eb", "d"],
    ], pl.List(pl.String)),
])
df.select(pl.col.a.list.sample(fraction=1.2))
```

Output:

```python
shape: (2, 1)
┌─────────────┐
│ a           │
│ ---         │
│ list[str]   │
╞═════════════╡
│ ["eb", "d"] │
└─────────────┘
```

The author of the original issue said this should raise an error because `fraction` is greater than one. As someone of Discord pointed out, if `fraction` is `1.6`, this **raises** an error. For example:

Input:

```python
import polars as pl
df = pl.DataFrame([
    pl.Series('a', [
        ["eb", "d"],
    ], pl.List(pl.String)),
])
df.select(pl.col.a.list.sample(fraction=1.6))
```

Output:

```python
Traceback (most recent call last):
  File "/Users/alex.gregory/Documents/test/main.py", line 7, in <module>
    print(df.select(pl.col.a.list.sample(fraction=1.6)))
          ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alex.gregory/Documents/test/.venv/lib/python3.13/site-packages/polars/dataframe/frame.py", line 10307, in select
    .collect(optimizations=QueryOptFlags._eager())
     ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alex.gregory/Documents/test/.venv/lib/python3.13/site-packages/polars/_utils/deprecation.py", line 97, in wrapper
    return function(*args, **kwargs)
  File "/Users/alex.gregory/Documents/test/.venv/lib/python3.13/site-packages/polars/lazyframe/opt_flags.py", line 326, in wrapper
    return function(*args, **kwargs)
  File "/Users/alex.gregory/Documents/test/.venv/lib/python3.13/site-packages/polars/lazyframe/frame.py", line 2440, in collect
    return wrap_df(ldf.collect(engine, callback))
                   ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
polars.exceptions.ShapeError: cannot take a larger sample than the total population when `with_replacement=false`
```

The exception is caused by the following line of code that computes `n` which is the number of items that will be sampled from the list:

```
let n = (s.as_ref().len() as f64 * fraction) as usize;
```

In the first example, `n` is `2` because the statement above is approximately `n = floor(2 * 1.2)` resulting in no error (floor comes from casting to `usize`). In the second example, `n` is `3` because the statement above is approximately `n = floor(2 * 1.6)`. Since `3` is greater than the size of the list, an error is raised.

This PR rounds up instead of rounding down. So, in the first example `n = ceil(2 * 1.2)` which is `3` and will result in an error. Does that make sense?

If I recall, because it has been a little while, the suggest was to not raise an error for this here because it's checked elsewhere thus reducing repetitive code.

**Comment by orlp** [comment]
I don't think the rounding behavior should be changed. There should just be a simple check to ensure the fraction is in [0, 1] early on.

**Comment by orlp** [comment]
Superseded by https://github.com/pola-rs/polars/pull/26964.
